Blish HUD/GameServices/Modules/Pkgs/GitHubPkgRepoProvider.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Flurl.Http;
+using Newtonsoft.Json;
+
+namespace Blish_HUD.Modules.Pkgs {
+    public class GitHubPkgRepoProvider : IPkgRepoProvider {
+
+        private static readonly Logger Logger = Logger.GetLogger<GitHubPkgRepoProvider>();
+
+        private const string GITHUB_RELEASES_URI = "/releases";
+        private const string ASSET_PACKAGE_NAME  = "packages.gz";
+
+        [Serializable]
+        public struct GitHubRelease {
+
+            public string Name { get; set; }
+
+            public GitHubAsset[] Assets { get; set; }
+
+        }
+
+        [Serializable]
+        public struct GitHubAsset {
+
+            public string Name { get; set; }
+
+            [JsonProperty("browser_download_url")]
+            public string BrowserDownloadUrl { get; set; }
+
+        }
+
+        protected virtual string RepoUrl { get; }
+
+        private static readonly Dictionary<string, PkgManifest[]> _pkgCache = new Dictionary<string, PkgManifest[]>();
+
+        protected GitHubPkgRepoProvider() { /* NOOP */ }
+
+        public GitHubPkgRepoProvider(string repoUrl) {
+            this.RepoUrl = repoUrl;
+        }
+
+        public async Task<bool> Load(IProgress<string> progress) {
+            if (!_pkgCache.ContainsKey(this.RepoUrl)) {
+                _pkgCache[this.RepoUrl] = await LoadRepo(progress);
+            }
+
+            return _pkgCache != null;
+        }
+
+        private async Task<PkgManifest[]> LoadRepo(IProgress<string> progress = null) {
+            progress?.Report(Strings.GameServices.ModulesService.PkgManagement_Progress_CheckingRepository);
+            var releaseResult = await UpdateGitHubReleases();
+
+            if (releaseResult.Exception == null) {
+                progress?.Report(Strings.GameServices.ModulesService.PkgManagement_Progress_GettingModuleList);
+                var manifests = await LoadPkgManifests(releaseResult.Releases);
+
+                if (manifests.Exception == null) {
+                    return manifests.PkgManifests;
+                } else {
+                    progress?.Report($"{Strings.GameServices.ModulesService.PkgManagement_Progress_FailedToReadOrParseRepoManifest}\r\n{manifests.Exception.Message}");
+                }
+            } else {
+                progress?.Report($"{Strings.GameServices.ModulesService.PkgManagement_Progress_FailedToGetReleases}\r\n{releaseResult.Exception.Message}");
+            }
+
+            return null;
```

Comment:
If you want, then you could flatten this construct by inverting all the ifs. IMO this would help readability.
e.g.
```cs
if (releaseResult.Exception != null) {
    progress?.Report("");
    return null;
}
...
if (manifests.Exception != null) {
    progress?.Report("");
    return null;
}

return manifests.PkgManifests;
```

Reply:
Having the same 
```cs
    progress?.Report("");
    return null;
```
written twice seems to really bother me.  I'll think it over / see how it looks.